src/widgets/modals/ByProgramModal.js

```diff
@@ -0,0 +1,406 @@
+};
+
+const getNewState = command => {
+  let newState;
+  switch (command) {
+    case 'RESET_ALL':
+      newState = {
+        program: {},
+        supplier: {},
+        period: {},
+        periods: null,
+        orderType: {},
+        orderTypes: null,
+        name: null,
+        currentStep: 0,
+      };
+      break;
+    case 'SELECT_PROGRAM':
+      newState = {
+        period: {},
+        periods: null,
+        supplier: {},
+        orderType: {},
+        orderTypes: null,
+        modalIsOpen: false,
+        currentStep: 1,
+      };
+      break;
+    case 'SELECT_ORDER_TYPE':
+      newState = { period: {}, periods: null, modalIsOpen: false };
+      break;
+    case 'SELECT_NAME':
+    case 'SELECT_PERIOD':
+    case 'SELECT_SUPPLIER':
+      newState = { modalIsOpen: false };
+      break;
+    default:
+      break;
+  }
+
+  return newState;
+};
+
+export class ByProgramModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      // Values used for deriving current UI state
+      currentStep: 0,
+      steps: null,
+      modalIsOpen: false,
+      isProgramBased: true,
+      // New requisition or stocktake values
+      program: {},
+      supplier: {},
+      orderType: {},
+      period: {},
+      name: null,
+      // Data stored for performance
+      storeTag: null,
+      programs: null,
+      suppliers: null,
+      orderTypes: null,
+      periods: null,
+    };
+  }
+
+  // Store unchanging select modal options and set the default
+  // steps array.
+  componentDidMount = () => {
+    const { settings, database } = this.props;
+    const programs = getAllPrograms(settings, database);
+    const suppliers = database.objects('Name').filtered('isSupplier = $0', true);
+    this.setCurrentSteps();
+    this.setState({ programs, suppliers });
+  };
+
+  getProgramValues = () => {
+    const { program, supplier, orderType, period, name } = this.state;
+    return { program, supplier, orderType, period, name };
+  };
+
+  getStepNumbers = () => {
+    const { isProgramBased } = this.state;
+    return {
+      program: 0,
+      supplier: isProgramBased ? 1 : 0,
+      orderType: 2,
+      period: 3,
+      name: isProgramBased ? 1 : 0,
+    };
+  };
+
+  // Open a modal - in case it is an earlier step in the process,
+  // set the current step to the step number corresponding with
+  // the onPress event.
+  openModal = ({ stepNumber }) => {
+    this.setState({ modalIsOpen: true, currentStep: stepNumber });
+  };
+
+  closeModal = () => {
+    this.setState({ modalIsOpen: false });
+  };
+
+  // Initial set up or update the current steps property to reflect the current state
+  setCurrentSteps = () => {
+    const sequentialSteps = this.getSequentialStepsProps();
+    const steps = this.getSteps().map(step => sequentialSteps[step]);
+    this.setState({ steps });
+  };
+
+  // Gets the available steps this modal can handle
+  getSteps = () => {
```

Comment:
I think it's neater to use objects here too:

return { 'requisition': 
             isProgramBased ? ['program', 'supplier', 'orderType', 'period'] : ['supplier'],
              'stocktake': 
             isProgramBased ? ['program', 'name'] : ['name'] 
           } [type]

Reply:
ended up removing this method as using an object for this made it unnecessary 3816a439a08b7de1910db19e0178c88558d06d45